src/Common/Hashing/PackageFileHasher.cs

```diff
@@ -0,0 +1,57 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System.Collections.Concurrent;
+using System;
+using System.IO;
+using BuildXL.Cache.ContentStore.Hashing;
+using System.Threading.Tasks;
+
+namespace Microsoft.MSBuildCache.Hashing;
+
+internal sealed class PackageFileHasher : IInputHasher
+{
+    private readonly ConcurrentDictionary<string, Lazy<Task<byte[]?>>> _cachedHashes = new(StringComparer.OrdinalIgnoreCase);
+
+    private readonly string _nugetPackageRoot;
+
+    private readonly IContentHasher _contentHasher;
+
+    public PackageFileHasher(string nugetPackageRoot, IContentHasher contentHasher)
+    {
+        _nugetPackageRoot = nugetPackageRoot;
+        _contentHasher = contentHasher;
+    }
+
+    public bool ContainsPath(string absolutePath) => absolutePath.IsUnderDirectory(_nugetPackageRoot);
+
+    public async ValueTask<byte[]?> GetHashAsync(string absolutePath)
+    {
+        if (!ContainsPath(absolutePath))
+        {
+            return null;
+        }
+
+        // Hashing is expensive so wrap it in a Lazy so that it happens *exactly* once.
+        Lazy<Task<byte[]?>> lazyTask = _cachedHashes.GetOrAdd(
+            absolutePath,
+            path => new Lazy<Task<byte[]?>>(
+                async () =>
+                {
+                    if (!File.Exists(path))
+                    {
+                        return null;
+                    }
+
+                    ContentHash contentHash;
+                    using (FileStream stream = File.OpenRead(path))
```

Comment:
>File.OpenRead(path)) [](http://example.com/codeflow?start=47&length=20)

Do we need to specify something for async?

Reply:
Added `FileOptions.Asynchronous | FileOptions.SequentialScan`. Hopefully the SequentialScan hint helps as well